### DISCUSSION TITLE
How to organize modes in subfolders

### DISCUSSION THREAD

**Discussion by bernarma** [missionpinball/missionpinball#151] [Q&A] [answered]
I have read the material (see link below) but haven't been able to create a number of related modes into a common sub-folder. I was wondering if anyone has been successful in doing this?

[(https://missionpinball.org/latest/game_logic/modes/#adding-your-modes-to-your-machine-configuration)]

My directory structure is:
> MPF/modes/mode_group
> MPF/modes/mode_group/mode1
> MPF/modes/mode_group/mode1/config
> MPF/modes/mode_group/mode2
> MPF/modes/mode_group/mode2/config

There are no files in the mode_group folder. I have also gone through the code and can't see how this has been implemented (unless I am looking at the wrong place).

#config_loader.py#

```def _load_modes(self, config_spec, machine_config, ignore_unknown_sections=False):```

**Comment by atummons** [comment]
What are you trying to do? All that is saying is that you can put them in folders for your personal organization. MPF will ignore folders without a config.yaml file.

**Comment by bernarma** [comment]
What I am trying to achieve is, instead of a large number of user-defined modes in the modes directory, I would like to group modes based on a specific group name and put them in there so that I can group them logically. What that translates to in practice is, instead of 30 modes at the top-level, I'll have maybe 6 folders at the top level which makes project navigation a little easier.

**Reply by atummons** [reply]
Then what you have shown is correct. In your config folders for the two modes listed you need a file (each) titled config.yaml Then in the main config folder you need to reference each mode by name in your modes: section.

**Reply by bernarma** [reply]
I tried that as that was my interpretation as well but in these two lines of code I can't see how that logic works and this logic is confirmed by raising an exception indicating that the mode is not found when I try it.

Both paths that are evaluated return (mpf_path or machine_path)/modes/mode1/config/mode1.config (always ignores the parent name which I am not using at all anywhere), I can't see how it would ever resolve it to the other file/directory structure.

```
            mpf_config_path = os.path.join(self.mpf_path, "modes", mode, 'config', mode + '.yaml')
            machine_config_path = os.path.join(self.machine_path, "modes", mode, 'config', mode + '.yaml')
            mode_config_files = []
            if os.path.isfile(mpf_config_path):
                self.log.debug("Loading mode %s from %s", mode, mpf_config_path)
                mode_config_files.append(mpf_config_path)
            if os.path.isfile(machine_config_path):
                self.log.debug("Loading mode %s from %s", mode, mpf_config_path)
                mode_config_files.append(machine_config_path)

            if not mode_config_files:
                _raise_mode_not_found_exception(mode)
```

**Reply by atummons** [reply]
You said mode1.config, but it should be mode1.yaml. Can you post a picture of your file explorer showing the file inside the config file with the full file tree visible? Also your modes: section from config.yaml.

**Reply by bernarma** [reply]
Sorry for the late reply. I made a mistake when I was describing the structure. I am using mode1.yaml (not config).

I am going to revisit this at a later stage as it was more related to the organizational structure as it isn't preventing me from doing anything at this stage.

**Comment by gbouthenot** [comment, marked as answer]
This feature have been removed in https://github.com/missionpinball/mpf/commit/49fb2c29cd29d0f5211d021942b464414e6dcf37 (refactor config loading into an injected class)
see, they were tests in `mpf/tests/test_Modes.py` and `mpf/tests/machine_files/mode_tests/config/test_modes.yaml` with "mode5, mode6, mode 7" as test modes. They were in sub_folders, and they were removed in this commit.
Sub folder are no longer searched for modes.
